components/brave_ads/content/browser/search_result_ad/search_result_ad_service.cc

```diff
@@ -343,55 +410,91 @@ AdsService* SearchResultAdService::SetAdsServiceForTesting(
   }
 
   if (!web_page) {
+    search_result_ads_map_[render_frame_host_id] = SearchResultAdsMap();
+    RunAdViewConfirmationPendingCallbacks(render_frame_host_id, false);
     return;
   }
 
-  SearchResultAdsList search_result_ads =
+  SearchResultAdsMap search_result_ads =
       ParseWebPageEntities(std::move(web_page));
 
-  std::reverse(search_result_ads.begin(), search_result_ads.end());
-  TriggerSearchResultAdViewedEvents(std::move(search_result_ads));
+  search_result_ads_map_.emplace(render_frame_host_id,
+                                 std::move(search_result_ads));
+
+  RunAdViewConfirmationPendingCallbacks(render_frame_host_id, true);
 }
 
-void SearchResultAdService::TriggerSearchResultAdViewedEvents(
-    SearchResultAdsList search_result_ads) {
-  DCHECK(ads_service_);
-  if (search_result_ads.empty()) {
-    return;
+void SearchResultAdService::RunAdViewConfirmationPendingCallbacks(
+    content::GlobalRenderFrameHostId render_frame_host_id,
+    bool ads_fetched) {
+  for (auto& callback_info :
+       view_confirmation_pending_callbacks_[render_frame_host_id]) {
+    bool ad_was_triggered = false;
+    if (ads_fetched) {
+      ad_was_triggered = TriggerAdViewConfirmation(
+          callback_info.creative_instance_id, render_frame_host_id);
+    }
+    if (ad_was_triggered) {
+      VLOG(1) << "Triggered search result ad viewed event for "
+              << callback_info.creative_instance_id;
+    } else {
+      VLOG(1) << "Failed to trigger search result ad viewed event for "
+              << callback_info.creative_instance_id;
+    }
+    std::move(callback_info.callback).Run(ad_was_triggered);
   }
+  view_confirmation_pending_callbacks_.erase(render_frame_host_id);
+}
+
+bool SearchResultAdService::TriggerAdViewConfirmation(
+    const std::string& creative_instance_id,
+    content::GlobalRenderFrameHostId render_frame_host_id) {
+  DCHECK(ads_service_);
 
-  ads::mojom::SearchResultAdPtr search_result_ad =
-      std::move(search_result_ads.back());
-  search_result_ads.pop_back();
+  SearchResultAdsMap& ads_map = search_result_ads_map_[render_frame_host_id];
+  auto it = ads_map.find(creative_instance_id);
+  if (it == ads_map.end()) {
+    return false;
+  }
 
+  ads::mojom::SearchResultAdPtr search_result_ad = std::move(it->second);
   ads_service_->TriggerSearchResultAdEvent(
       std::move(search_result_ad), ads::mojom::SearchResultAdEventType::kViewed,
       base::BindOnce(
           &SearchResultAdService::OnTriggerSearchResultAdViewedEvents,
-          weak_factory_.GetWeakPtr(), std::move(search_result_ads)));
+          weak_factory_.GetWeakPtr()));
+
+  ads_map.erase(creative_instance_id);
+
+  return true;
 }
 
 void SearchResultAdService::OnTriggerSearchResultAdViewedEvents(
-    SearchResultAdsList search_result_ads,
     const bool success,
     const std::string& placement_id,
     ads::mojom::SearchResultAdEventType ad_event_type) {
```

Comment:
nit: remove as unused or make the function generic for any event type, i.e. `OnDidTriggerSearchResultAdEvents`